File: partiql-plan/src/main/java/org/partiql/plan/rel/RelProject.java
package org.partiql.plan.rel;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operator;
import org.partiql.plan.OperatorVisitor;
import org.partiql.plan.rex.Rex;

import java.util.List;

/**
 * Logical project abstract base class.
 */
public abstract class RelProject extends RelBase {

    /**
     * @return new {@link RelProject} instance
     */
    @NotNull
    public static RelProject create(Rel input, List<Rex> projections) {
        return new Impl(input, projections);
    }

    /**
     * @return input rel (operand 0)
     */
    @NotNull
    public abstract Rel getInput();

    /**
     * @return projection (not a operand, it's a list not an operator).
     */
    @NotNull
    public abstract List<Rex> getProjections();

    @NotNull
    @Override
    protected final RelType type() {
        throw new UnsupportedOperationException("Derive type is not implemented");
    }

    @NotNull
    @Override
    protected final List<Operator> operands() {
        Rel c0 = getInput();
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(@NotNull OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitProject(this, ctx);
    }

    @NotNull
    public abstract RelProject copy(@NotNull Rel input);

    @NotNull
    public abstract RelProject copy(@NotNull Rel input, @NotNull List<Rex> projections);

    private static class Impl extends RelProject {

        private final Rel input;
        private final List<Rex> projections;

        private Impl(Rel input, List<Rex> projections) {
            this.input = input;
            this.projections = projections;
        }

        @NotNull
        @Override
        public Rel getInput() {
            return input;
        }

        @NotNull
        @Override
        public List<Rex> getProjections() {
            return projections;
        }

        @NotNull
        @Override
        public RelProject copy(@NotNull Rel input) {
            return new Impl(input, projections);
        }

        @NotNull
        @Override
        public RelProject copy(@NotNull Rel input, @NotNull List<Rex> projections) {
            return new Impl(input, projections);
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rel/RelProject.kt
package org.partiql.plan.v1.rel

import org.partiql.plan.v1.rex.Rex

/**
 * Logical `PROJECTION` operator
 */
public interface RelProject : Rel {

    public fun getInput(): Rel

    public fun getProjections(): List<Rex>

    override fun getInputs(): List<Rel> = listOf(getInput())

    override fun isOrdered(): Boolean = getInput().isOrdered()

    public override fun <R, C> accept(visitor: RelVisitor<R, C>, ctx: C): R = visitor.visitProject(this, ctx)

    /**
     * Default [RelProject] implementation meant for extension.
     */
    public abstract class Base(input: Rel, projections: List<Rex>) : RelProject {

        // DO NOT USE FINAL
        private var _input = input
        private var _projections = projections

        private var _inputs: List<Rel>? = null

        override fun getInput(): Rel = _input

        override fun getProjections(): List<Rex> = _projections

        override fun getInputs(): List<Rel> {
            if (_inputs == null) {
                _inputs = listOf(_input)
            }
            return _inputs!!
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is RelProject) return false
            if (_input != other.getInput()) return false
            if (_projections != other.getProjections()) return false
            return true
        }

        override fun hashCode(): Int {
            var result = 1
            result = 31 * result + _input.hashCode()
            result = 31 * result + _projections.hashCode()
            return result
        }
    }
}
